### PR TITLE
MBS-12251 [Instrumentation test plugin] use new AGP variants API

### PR DESCRIPTION
**Pull request by dsvoronin**
Gradle plugin refactoring:
- use new variant API for AGP
- use "configurators" to set groups of task properties
- add LocalRunInteractor to extract part of the plugin that works only to set correct args for local runs (should be probably removed in future)

~~- [ ] [fastCheck build](http://links.k.avito.ru/6Y)~~
- [x] [fastCheck build](http://links.k.avito.ru/HY)

### CODE REVIEW

File: subprojects/gradle/cd/src/gradleTest/kotlin/com/avito/ci/steps/UploadCdBuildResultIntegrationTest.kt [hunk 1]
```diff
@@ -166,7 +167,9 @@ class RealTest {
             ).tryParseRev("HEAD").getOrThrow()
         )
 
-        val runId = "runId"
+        val commit = projectDir.git("rev-parse HEAD").trim()
+
+        val runId = "$commit.teamcity-BT"
```
Comment (dsvoronin) [inline subprojects/gradle/cd/src/gradleTest/kotlin/com/avito/ci/steps/UploadCdBuildResultIntegrationTest.kt:172]: behavior change: `runId` set from `build.gradle` config `android.defaultConfig.testInstrumentationArgs` will be ignored
reason: we build complex RunId to be correct for our report engine, overriding here can be problematic

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/AndroidVariantConfiguratorFactory.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package com.avito.instrumentation.internal
+
+import com.android.build.api.variant.ApplicationVariant
+import com.android.build.api.variant.LibraryVariant
+import com.android.build.api.variant.Variant
+
+internal class AndroidVariantConfiguratorFactory {
+
+    fun createConfigurator(variant: Variant): AndroidVariantConfigurator<*>? {
```
Comment (sergioboi) [inline subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/AndroidVariantConfiguratorFactory.kt:9]: Strange create that could return null

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/BuildEnvResolver.kt [hunk 3]
```diff
@@ -3,13 +3,13 @@ package com.avito.instrumentation.internal
 import com.avito.utils.gradle.envArgs
 import org.gradle.api.Project
 
-internal object BuildEnvResolver {
+internal class BuildEnvResolver(private val project: Project) {
```
Comment (sergioboi) [inline subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/BuildEnvResolver.kt:6]: I'm not sure why we need a project reference there

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ConfiguratorsFactory.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    val localRunInteractor: LocalRunInteractor = LocalRunInteractor(
+        loggerFactory = loggerFactory,
+        argsTester = argsTester,
+        dslInteractor = androidDslInteractor
+    )
+
+    fun createTaskConfigurators(
+        configuration: InstrumentationConfiguration,
+        variant: Variant
+    ): List<InstrumentationTaskConfigurator>? {
```
Comment (sergioboi) [inline subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ConfiguratorsFactory.kt:69]: emptylist?

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/GitResolver.kt [hunk 5]
```diff
@@ -4,17 +4,17 @@ import com.avito.git.gitState
 import org.gradle.api.Project
 import org.gradle.api.provider.Provider
 
-internal object GitResolver {
+internal class GitResolver(private val project: Project) {
```
Comment (sergioboi) [inline subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/GitResolver.kt:7]: The same, too wide scope to save whole Project there

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/InstrumentationTaskConfigurator.kt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.avito.instrumentation.internal
+
+import com.avito.instrumentation.InstrumentationTestsTask
+
+internal interface InstrumentationTaskConfigurator {
+
+    fun configure(task: InstrumentationTestsTask)
```
Comment (sergioboi) [inline subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/InstrumentationTaskConfigurator.kt:7]:
```suggestion
    fun configure(task: Provider<InstrumentationTestsTask>)
```
Would be more Gradle way

Comment (sergioboi) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/InstrumentationTaskConfigurator.kt:7]: Or you guarantee laziness at the upper level?

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ReportResolver.kt [hunk 7]
```diff
@@ -3,34 +3,35 @@ package com.avito.instrumentation.internal
...
 import com.avito.instrumentation.configuration.ReportViewer
 
-internal class ReportResolver(private val runIdResolver: RunIdResolver) {
+internal class ReportResolver(
+    private val extension: InstrumentationTestsPluginExtension,
+    private val runIdResolver: RunIdResolver
+) {
 
-    fun getReportApiUrl(extension: InstrumentationTestsPluginExtension): String {
+    fun getReportApiUrl(): String {
```
Comment (sergioboi) [inline subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ReportResolver.kt:11]:
```suggestion
    val reportApiUrl: String
```
Why we don't use the property?

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/AndroidVariantConfigurator.kt [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package com.avito.instrumentation.internal
+
+import com.android.build.api.variant.Variant
+
+internal interface AndroidVariantConfigurator<T : Variant> : InstrumentationTaskConfigurator {
+
+    val variant: T
```
Comment (sergioboi) [inline subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/AndroidVariantConfigurator.kt:7]: I don't get why we need the property there
Looks like an abstract class with constructor argument

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/AndroidVariantConfiguratorFactory.kt (hunk 2 above)
Comment (dsvoronin) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/AndroidVariantConfiguratorFactory.kt:9]: yep, i'm thinking of improvement, 
see here https://github.com/avito-tech/avito-android/pull/1261/files#diff-a9f5557634124eb2a0ba2c4457392e08c1c8ead18a370ff37347d311470e210cR48

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/BuildEnvResolver.kt (hunk 3 above)
Comment (dsvoronin) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/BuildEnvResolver.kt:6]: You mean to pass `envArgs`, or even `build` directly?

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ConfiguratorsFactory.kt (hunk 4 above)
Comment (dsvoronin) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ConfiguratorsFactory.kt:69]: should work, but problem lies here https://github.com/avito-tech/avito-android/pull/1261#discussion_r738319406

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/GitResolver.kt (hunk 5 above)
Comment (dsvoronin) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/GitResolver.kt:7]: ok, i'll pass `gitState` directly

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/InstrumentationTaskConfigurator.kt (hunk 6 above)
Comment (dsvoronin) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/InstrumentationTaskConfigurator.kt:7]: I couldn't find a way to lazily configure task at the moment of registration, like here:
https://github.com/avito-tech/avito-android/pull/1261/files#diff-a9f5557634124eb2a0ba2c4457392e08c1c8ead18a370ff37347d311470e210cR57

It wants Action<InstrumentationTestsTask> as parameter, which i use but only to iterate through `Configurators`
Not sure if passing knowledge that it is a configuration is worth it.

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ReportResolver.kt (hunk 7 above)
Comment (dsvoronin) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ReportResolver.kt:11]: i think we can

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/AndroidVariantConfigurator.kt (hunk 8 above)
Comment (dsvoronin) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/AndroidVariantConfigurator.kt:7]: abstract class should work, though about more resposibility for this configurator, like more interfaces applied, but ended up single inheritance

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/BuildEnvResolver.kt (hunk 3 above)
Comment (dsvoronin) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/BuildEnvResolver.kt:6]: can't in fact: envArgs should be lazy, it's a debt: https://github.com/avito-tech/avito-android/blob/71c71f8eb633b4e4bfc81984fe3f62b967b3806a/subprojects/gradle/gradle-extensions/src/main/kotlin/com/avito/kotlin/dsl/ProjectExtensions.kt#L92

File: subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ReportResolver.kt (hunk 7 above)
Comment (dsvoronin) [reply subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/ReportResolver.kt:11]: ReportViewer model is not a properties, just var's https://github.com/avito-tech/avito-android/blob/71c71f8eb633b4e4bfc81984fe3f62b967b3806a/subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/InstrumentationTestsReportExtension.kt#L21

so it should be queried lazily, properties with getters will work, but not with eager initializers

**Pull request merged**